script/src/verify.rs

```diff
@@ -455,6 +467,267 @@ impl<'a, DL: CellDataProvider + HeaderProvider> TransactionScriptsVerifier<'a, D
         Ok(cycles)
     }
 
+    fn make_snapshot(
+        &self,
+        mut vm: Machine<'a>,
+        group: &ScriptGroup,
+        current: (ScriptGroupType, Byte32),
+        remain: Vec<(ScriptGroupType, Byte32)>,
+        current_cycles: Cycle,
+        limit_cycles: Cycle,
+    ) -> Result<TransactionSnapshot, Error> {
+        let snap = make_snapshot(&mut vm.machine).map_err::<Error, _>(|e| {
+            ScriptError::VMInternalError(format!("{:?}", e))
+                .source(group)
+                .into()
+        })?;
+
+        Ok(TransactionSnapshot {
+            current,
+            remain,
+            snap,
+            current_cycles,
+            limit_cycles,
+        })
+    }
+
+    /// Performing a resumable verification on the transaction scripts.
+    ///
+    /// ## Params
+    ///
+    /// * `limit_cycles` - Maximium allowed cycles to run the scripts. The verification quits early
+    /// when the consumed cycles exceed the limit.
+    ///
+    /// ## Returns
+    ///
+    /// It returns the total consumed cycles if verification completed,
+    /// If verify is suspended, a snapshot will retruned.
+    pub fn resumable_verify(&self, limit_cycles: Cycle) -> Result<VerifyResult, Error> {
+        let mut cycles = 0;
+
+        let groups: Vec<_> = self.groups().collect();
+        for (idx, (ty, hash, group)) in groups.iter().enumerate() {
+            // vm should early return invalid cycles
+            let step_cycle = limit_cycles.checked_sub(cycles).ok_or_else(|| {
+                ScriptError::VMInternalError(format!(
+                    "expect invalid cycles {} {}",
+                    limit_cycles, cycles
+                ))
+                .source(group)
+            })?;
+
+            match self.verify_group_with_chunk(&group, step_cycle, None) {
+                Ok(ChunkState::Completed(cycle)) => {
+                    cycles = wrapping_cycles_add(cycles, cycle, limit_cycles, group)?;
```

Comment:
I personally think that the cycles add/reduce check here is unnecessary, because vm uses checked add when calculating cycles, and it must not be greater than limit